### PR TITLE
Editorial: Convert @@ notation to intrinsics notation for well-known Symbols

### PR DESCRIPTION
**Pull request by ljharb**
- this avoids a conceptual conflict with decorators, when they land

Open to bikesheds; other alternatives I considered:
 - explicitly using intrinsics, ie, `%SymbolSpecies%`, and hardcoding that table, while removing the "well-known symbols table". This felt like it would lose some clarity around intrinsics vs well-known-symbols.
 - Using `%Symbol%species%` - this would require upstream support in ecmarkup, so i opted not to do it for now.
 - something else?

### CODE REVIEW

**Comment by gibson042** [comment]
Will this interact with special treatment of `%%` by ecmarkup (e.g., for "%%Internal%%" in [spec](https://github.com/bterlson/ecmarkup/blob/a5ce61c6a8bdbd28e9c087760d450e58927b880c/test/algorithms.html) → [HTML](https://github.com/bterlson/ecmarkup/blob/a5ce61c6a8bdbd28e9c087760d450e58927b880c/test/baselines/reference/algorithms.html))? And as a meta question, do we want to _introduce_ auto-linking of well-known symbols if we're changing their representation?

**Comment by ljharb** [comment]
@gibson042 i would love to see auto-linking of well-known symbols, with or without this change.

I've tested the HTML output locally and atm they don't link anywhere, before or after.

**Comment by annevk** [comment]
This would also affect the HTML Standard.

cc @domenic

**Comment by littledan** [comment]
Would it be possible to use a syntax more similar to `Symbol.foo`, so that it looks more similar to how JavaScript developers refer to the same thing?

**Comment by ljharb** [comment]
What’d you have in mind? It would have to be apparent that it’s not a property lookup on %Symbol%, i suspect.

**Comment by littledan** [comment]
Sorry, I don't have a concrete suggestion, and agree it shouldn't imply actual property access. But I would prefer that editorial churn like this would somehow inch us closer to being readable by normal programmers. `@@` is now understood by a small set of people outside TC39; using a new syntax will make it harder to understand for that see while keeping it unintuitive for the big majority.

**Comment by bakkot** [comment]
My preference would be to use the %SymbolSpecies% notation and move the well-known symbols table to immediately following the intrinsics table. That way we reduce the total number of novel syntaxes in the spec. I don't actually think there is much of an important distinction between intrinsics and well-known symbols except for the cross-realm interaction. They're similar enough that it makes sense to put them near each other and use the same syntax, IMO.

**Comment by ljharb** [comment]
@bakkot i initially looked at doing that, but it seemed like a loss to remove the useful info in this [section](https://tc39.github.io/ecma262/#sec-well-known-symbols).

**Comment by bakkot** [comment]
[figure-2.png](https://github.com/tc39/ecma262/blob/7acacc524213058a2368b5fa751fac7ea08ba230/img/figure-2.png) also needs to be updated, since it contains the text `@@hasInstance`.

**Comment by zenparsing** [comment]
I'm thinking that we should leave things the way they are for now, since `@@` is well-understood, fairly readable, and used by other specs.

**Comment by michaelficarra** [comment]
Decision from the editor call today was to change `@@symbolName` notation to `%Symbol.symbolName%` for consistency with other instrinsics.

**Comment by michaelficarra** [comment]
@ljharb Don't forget @bakkot's comment about the usage in figures: https://github.com/tc39/ecma262/pull/1314#issuecomment-441111136

**Comment by bakkot** [comment]
Also would be good to have a NOTE near the symbol table mentioning the old notation, I think.

**Comment by jmdyck** [comment]
In this PR, do well-known symbols qualify as intrinsics?

- If they do, that would require some changes to [6.1.7.4 Well-Known Intrinsic Objects](https://tc39.es/ecma262/#sec-well-known-intrinsic-objects), which indicates that intrinsics (or at least, the well-known ones) are all objects.

- If they don't, then the preamble to the "Well-Known Symbols" table shouldn't say that it lists intrinsics.

Either way, it would be nice if the definition of the %X.Y% notation didn't say that it means an object.

(For reference, here are some other discussions about the definition of _intrinsics_:
- [how-we-work PR #64](https://github.com/tc39/how-we-work/pull/64)
- [comment](https://github.com/tc39/ecma262/pull/1376#issuecomment-492257182) in PR #1376 
- issue #1540 
- issue #1703 

)

**Comment by bakkot** [comment]
I think it's easier if we don't try to make them into intrinsics. Apart from being primitives, they also do not have different identities across realms. They're really a different kin of thing.

> If they don't, then the preamble to the "Well-Known Symbols" table shouldn't say that it lists intrinsics.

Agreed. I think it needn't mention that at all: just "Within this specification a well-known symbol is referred to using the standard intrinsic notation", possibly followed by "defined in section 6.1.7.4", is probably sufficient.

---

> Either way, it would be nice if the definition of the %X.Y% notation didn't say that it means an object.

Where does it say that? The only relevant text I see says

> Within this specification a reference such as %name% means the intrinsic object, associated with the current realm, corresponding to the name. A reference such as %name.a.b% means, as if the "b" property of the "a" property of the intrinsic object %name% was accessed prior to any ECMAScript code being evaluated.

Which indicates that %X% is an object (which will still be the case after this PR), but does not say that %X.Y% must be.

**Comment by allenwb** [comment]
This is exactly why two distinct notations were used in the first place. Well known symbols are not intrinsics and the distinction is emphasized by using two notations.

This change seems like pointless churm that in the end blurs the important semantic distinctions.

**Comment by jmdyck** [comment]
> > Either way, it would be nice if the definition of the %X.Y% notation didn't say that it means an object.
> 
> Where does it say that?

Whoops, right. I saw "means the intrinsic object" and failed to see that it only applied to the dotless form.

**Comment by zenparsing** [comment]
I agree with Allen's comment and think that, in general, the use of `%X.Y%` does not result in an overall increase in spec clarity.

**Comment by ljharb** [comment]
Well-known symbols are nonconfigurable nonwritable properties on an intrinsic; what's the important semantic distinction?

@bakkot @michaelficarra note added and image updated.

**Comment by michaelficarra** [comment]
@ljharb Any reason this PR is a draft currently? It LGTM.

**Comment by ljharb** [comment]
@michaelficarra the previous feeling on it had been that it would cause too much churn for not enough gain. Let's chat about it next week!

**Comment by michaelficarra** [comment]
@ljharb I'm happy to talk about it more if you want, but we already made the call to make this change: https://github.com/tc39/ecma262/pull/1314#issuecomment-556553406. Has something changed since then?

**Comment by ljharb** [comment]
oh oops :-) i guess i forgot/missed it. In that case, I’ll convert it back, and rebase!

**Comment by bakkot** [comment]
Eh, revisiting this since the last time it was discussed in light of the significant amount of other nontrivial churn we've made recently or propose to make soon (#1914, #1964, #1796, #1950, #1889, #1595, etc), my preference is now to hold off on this until we've dealt with some of the higher-value stuff and let things rest for a while.

**Comment by ljharb** [comment]
That’s fine too; but the work is already done and won’t likely impact those other PRs much (and I’m happy to rebase them all myself after this lands, if that’s a concern)

**Comment by bakkot** [comment]
I am less concerned with rebasing and more concerned with the amount of change that readers of the spec need to keep up with over time.

**Comment by littledan** [comment]
Personally, I'm really positive about this change. The @@ syntax has caused some confusing spill-over into documentation and design discussions, and it really doesn't make sense given that ES6 didn't end up including that syntax in the language. This PR adds a lot of clarity, and can help guide others towards communicating with clarity. Good work, @ljharb .

File: spec.html
Comment (ExE-Boss) [inline spec.html:26544]: These should have `<dfn>` on them, otherwise auto‑linking won’t work:
```suggestion
        <p>The initial value of `Symbol.asyncIterator` is the well known symbol <dfn>%Symbol.asyncIterator%</dfn> (<emu-xref href="#table-1"></emu-xref>).</p>
```

File: spec.html
Comment (ExE-Boss) [inline spec.html:40860]: This should have a more specific ID that is unlikely to conflict:

```suggestion
    <emu-clause oldids="sec-@@tostringtag" id="sec-module-namespace-objects-tostringtag">
```
or, depending on which section ID style gets chosen:
```suggestion
    <emu-clause oldids="sec-@@tostringtag" id="sec-module-namespace-objects-%symbol.tostringtag%">
```

File: spec.html
Comment (ExE-Boss) [inline spec.html:40861]: This should probably be wrapped in brackets:

```suggestion
      <h1>[ %Symbol.toStringTag% ]</h1>
```

File: spec.html
Comment (ljharb) [reply spec.html:26544]: It will work without dfns once ecmarkup is modified to accept the new syntax, which it will be before merging this PR.

File: spec.html
Comment (ljharb) [reply spec.html:40860]: it won’t conflict with anything else because there will never be another top level section for this value

File: spec.html
Comment (ExE-Boss) [reply spec.html:26544]: **ecmarkup** already supports this syntax, since `%TypedArray.prototype%` links to <https://tc39.es/ecma262/#sec-properties-of-the-%typedarrayprototype%-object>.

Comment (ljharb) [reply spec.html:26544]: In that case i think intrinsics should all not require the dfn, unless all the rest of them have it?

Comment (ExE-Boss) [reply spec.html:26544]: All the rest of the intrinsics have `<dfn>`s, as it’s necessary for auto‑linking.

File: spec.html
Comment (ExE-Boss) [reply spec.html:40860]: Well, I still feel like we should fix the section ID.

Comment (ljharb) [reply spec.html:40860]: I don't see the value of that; `sec-%symbol.tostringtag%` isn't going to ever conflict with anything else. this change just makes the URL longer for no value.

File: spec.html
Comment (ljharb) [reply spec.html:26544]: fair enough, then these should too, thanks :-)

Comment (ExE-Boss) [reply spec.html:26544]: <https://github.com/tc39/ecma262/pull/1996> has now landed, so this needs to be rebased on top of it.

File: spec.html
Comment (ljharb) [reply spec.html:40861]: That's not the current convention, or the previous text; the property key is a Symbol, which is what `%Symbol.toStringTag%` is.

**Comment by ljharb** [comment]
@ExE-Boss The async generator style was a mistake that hasn't been worthwhile yet to correct; the former style is the reigning convention.

File: spec.html
Comment (ExE-Boss) [inline spec.html:40773]:
```suggestion
      <emu-clause id="sec-asyncfunction.prototype-%symbol.tostringtag%" oldids="sec-async-function-prototype-properties-toStringTag">

File: spec.html
Comment (ExE-Boss) [inline spec.html:39226]:
```suggestion
      <emu-clause id="sec-asyncgeneratorfunction.prototype-%symbol.tostringtag%" oldids="sec-asyncgeneratorfunction-prototype-tostringtag">
```

**Comment by ljharb** [comment]
Why does the order matter?

**Comment by ExE-Boss** [comment]
For consistency with the rest of the specification.

**Comment by ljharb** [comment]
Right but why does this consistency matter? the only thing that ends up in the rendered output is `id`.

**Comment by bakkot** [comment]
Per discussion today, let's do this, but add a sentence for each symbol somewhere (probably under each [Symbol.whatever subclause](https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-symbol.hasinstance)) to the effect of "In previous versions of this specification, this was referred to as <dfn>@@hasInstance</dfn>".

**Comment by ljharb** [comment]
PR's rebased.

> but add a sentence for each symbol somewhere

it seems a bit awkward to add it there, but I can - there's already a sentence in the well-known symbols table ```Previous editions of this specification used a notation of the form @@name, where the current edition would use `%Symbol.name%` ```, but if it's important that each former form be directly greppable, then it could also be another column in the well-known symbols table. Thoughts?

**Comment by bakkot** [comment]
I would like the old names to be greppable, yeah. I'm fine going with a column in that table ("previous name" or whatever) if the other editors are.

**Comment by ljharb** [comment]
added it as a separate commit, so i can drop it easily if needed.

**Comment by michaelficarra** [comment]
@bakkot I don't think we need to mention every single old name. If someone wants to know what happened to `@@species` and find-in-page for `@@species` doesn't turn anything up, find-in-page for `@@` will bring them to the explanation of what happened. Similarly, find-in-page for `species` will also make them aware of the notation change. That seems sufficient.

**Comment by Josh-Cena** [comment]
MDN (and probably other developer references) use the `@@` notation, so hard to say this is a spec-only editorial change. I'm not against the idea in general but I'd admit it makes text much longer.

**Comment by michaelficarra** [comment]
@Josh-Cena Brevity isn't that high on our list of editorial concerns.

**Comment by Josh-Cena** [comment]
I understand it's not a concern for the spec, just wanted to share my opinions as the MDN editor supposing I received a change like this—because this will certainly have an impact on us. Not sure if the community churn of old browser release notes, blog posts, and developer references have been taken into account in the decision making.

**Comment by zloirock** [comment]
@michaelficarra It's not about brevity. It affects a huge amount of documentation - many projects I participate in use `@@` notation. A huge amount of archive discussions where it's used will become more confusing for beginners.

**Comment by ljharb** [comment]
@Josh-Cena the feedback i'd have given to MDN authors back in 2015 (or to anyone writing documentation) would have been to staunchly avoid using spec-internal notation, and to just use `Symbol.whatever` - which is what I'd suggest changing MDN to regardless, before or after this PR. It's certainly a lot of churn, which is unfortunate.

**Comment by Josh-Cena** [comment]
I don't have strong opinions about changing it for MDN for the future, just wondering if the editors have thought about it for materials from the past.

So would you recommend `Array.prototype[Symbol.iterator]` or `Array.prototype[%Symbol.iterator%]`, for both pedagogical and pedantic concerns?

**Comment by ljharb** [comment]
The former, exclusively. Spec-internal notation may change at any time based on the whims of the editors, and shouldn’t be relied upon externally. I’ve made that same mistake in a few of my packages.

**Comment by Josh-Cena** [comment]
`%Symbol.iterator%` is very, very unlikely to change, and may even be part of the language with `getIntrinsic` though.

**Comment by ljharb** [comment]
@Josh-Cena nope, the getIntrinsic proposal (which I'm championing) explicitly avoids exposing the percent signs to users, for the reasons indicated.

**Comment by Josh-Cena** [comment]
> explicitly avoids exposing the percent signs to users

Really? That's not the impression I get from either the README or the spec:

> ```js
> const $String = getIntrinsic('%String%');
> ```
>
> Provide a new global or static function, `getIntrinsic`, that takes a string argument and effectively replicates the `get-intrinsic` package API, as well as the specification's internal `%a.b.c%` notation. It would return the "original" value being requested.

Was there a recent API change?

**Comment by ljharb** [comment]
Looks like I just forgot to update the readme, I'll do so now :-)

**Comment by ljharb** [comment]
Updated text per editor call; this is now only blocked by editor review, and ecmarkup supporting Symbol intrinsics in AO h1s.

**Comment by Jack-Works** [comment]
What about things like `%RegExp.prototype[@@match]%`? Does it become `%RegExp.prototype[%Symbol.match%]%`?

**Comment by ljharb** [comment]
Yep, exactly.

File: spec.html
Comment (syg) [inline spec.html:1207]:
```suggestion
        <emu-note>Previous editions of this specification used a notation of the form @@name, where the current edition would use `%Symbol.name%`. In particular, the following names were used: @@asyncIterator, @@hasInstance, @@isConcatSpreadable, @@iterator, @@match, @@matchAll, @@replace, @@search, @@species, @@split, @@toPrimitive, @@toStringTag, and @@unscopables.</emu-note>
```

**Review by syg — APPROVED** [review]
lgtm % wording

**Comment by bakkot** [comment]
OK, if this is rebased and updated to incorporate ecmarkup 19.0.0 it should be ready to go.

**Comment by ljharb** [comment]
k, should be all rebased and up to date, and the SVG updated (i had to move some of the coordinates around to make the longer text fit nicely)

File: spec.html
Comment (jmdyck) [inline spec.html:30620]: Reinstate hyphen, just for consistency:
```suggestion
        <p>The initial value of `Symbol.asyncIterator` is the well-known symbol %Symbol.asyncIterator% (<emu-xref href="#table-well-known-symbols"></emu-xref>).</p>
```

**Comment by zloirock** [comment]
I still hope that it will not be merged for [this reason](https://github.com/tc39/ecma262/pull/1314#issuecomment-1728529600).

File: img/figure-2.svg
Comment (Josh-Cena) [inline img/figure-2.svg:1]: FYI: the object names are no longer centered, and the "Function" -> "Function.prototype" arrow is not sticking to the edge of the box.

**Comment by ljharb** [comment]
https://github.com/tc39/ecma262/pull/1314#issuecomment-1728535369

> the feedback i'd have given to MDN authors back in 2015 (or to anyone writing documentation) would have been to staunchly avoid using spec-internal notation, and to just use `Symbol.whatever` - which is what I'd suggest changing MDN to regardless, before or after this PR. It's certainly a lot of churn, which is unfortunate.

**Comment by Josh-Cena** [comment]
Lots of historical materials will stick around and will confuse readers. Again I'm willing to do the work so MDN aligns with the spec, but the ship has sailed (as tc39 always says).

**Comment by ljharb** [comment]
True enough, but that's often unavoidable, and in this case, it'll be far more confusing if we have a spec with both decorators and `@@` in it.

**Comment by Josh-Cena** [comment]
Decorators don't use `@@` though, and code entities are typeset differently in the spec, if they are used at all. I see it can be confusing for MDN (which is another thing to think about, for us!), but not for the spec.

**Review by jmdyck — CHANGES_REQUESTED** [review]

File: spec.html
Comment (jmdyck) [inline spec.html:30844]: Accidental revert of PR #2592?
```suggestion
        <p>This method is called by ECMAScript language operators to convert a Symbol object to a primitive value.</p>
```

**Review by jmdyck — COMMENTED** [review]
You gave a new clause id to every intrinsic with the property key `%Symbol.foo%`, except for:
- `RegExp.prototype [ %Symbol.matchAll% ]`
- `%AsyncIteratorPrototype% [ %Symbol.asyncIterator% ]`
- `AsyncFunction.prototype [ %Symbol.toStringTag% ]`

Did you intend to skip those?

Apart from that, looks okay to me.

**Comment by ljharb** [comment]
I didn't do it intentionally, it's just that I only changed the ones that had `@@` in them.

I could change them to be consistent, but I'm not sure if that's worth the churn + oldids.

**Comment by jmdyck** [comment]
> I didn't do it intentionally, it's just that I only changed the ones that had `@@` in them.

That was my guess at first, but then I noticed that you had introduced new ids to replace
- `sec-asyncgeneratorfunction-prototype-tostringtag`
- `sec-asyncgenerator-prototype-tostringtag`

which don't have `@@`.

**Comment by ljharb** [comment]
Guess I was just being inconsistent in my application then :-) I'll update them to match.

**Review by michaelficarra — APPROVED** [review]
Still LGTM.

File: spec.html
Comment (jmdyck) [inline spec.html:46402]:
```suggestion
      <emu-clause oldids="sec-asynciteratorprototype-asynciterator" id="sec-%asynciteratorprototype%-%symbol.asynciterator%">
```
because we usually include the base object's percents in the id. (The status quo id is anomalous on this point.)

Comment (michaelficarra) [reply spec.html:46402]: Either one WFM. I'm not too worried about consistency in clause IDs.

**Comment by Josh-Cena** [comment]
@ljharb Out of curiosity: what's up with "luke_skywalker"?

**Comment by ljharb** [comment]
@Josh-Cena this PR kills at-at's

**Comment by Josh-Cena** [comment]
Oh! lmao 😄

**Comment by domenic** [comment]
Just want to drop in with an extra thanks to @ljharb for doing the HTML and Web IDL PRs. That kind of cross-ecosystem citizenship is much appreciated.

**Comment by linusg** [comment]
PR for ECMA-402: https://github.com/tc39/ecma402/pull/905

**Comment by ljharb** [comment]
whoops, i should have made that one too. thanks!

**Pull request merged**
